<a href="https://colab.research.google.com/github/ikwak2/StatBook/blob/main/ch10/10_5_R_prac.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 10.5 R 연습

## <예제 10-2>의 자료에 대한 분산분석 과정

### 자료 입력

[Note] R에서 csv 파일 불러오기
R에서 Excel(xls,xlsx) 파일을 불러오기 위해서는 특별한 패키지가 필요하므로, csv  파일로 저장하여 read.csv 명령어를 이용하여 불러온다.

In [1]:
## 현재 폴더로 예제파일 ex10_2.csv , ex10_4.csv 파일을 다운받습니다. 
system("wget https://raw.githubusercontent.com/ikwak2/StatBook/main/ch10/ex10_2.csv")
system("wget https://raw.githubusercontent.com/ikwak2/StatBook/main/ch10/ex10_4.csv")

In [2]:
ex.10<-read.csv("ex10_2.csv",header=T)
ex.10

value method
1  65    A     
2  87    A     
3  73    A     
4  79    A     
5  81    A     
6  69    A     
7  75    B     
8  69    B     
9  83    B     
10 81    B     
11 72    B     
12 79    B     
13 90    B     
14 59    C     
15 78    C     
16 67    C     
17 62    C     
18 83    C     
19 76    C     
20 94    D     
21 89    D     
22 80    D     
23 88    D

read.csv 명령어를 이용하여 ex10_2 파일을 ex.10에 저장한다. 첫 번째 변수 value는 시험성적, 두 번째 변수 method는 교육방법이다. 변수명 및 데이터경로에 한글이 있을 경우 오류가 생길 수 있으므로, 가능하면 영문을 사용하는 것이 좋다.  

### 일원분류 분산분석

In [3]:
ex.10.aov<-aov(value~method,data=ex.10)
ex.10.aov

Call:
   aov(formula = value ~ method, data = ex.10)

Terms:
                   method Residuals
Sum of Squares   712.5864 1196.6310
Deg. of Freedom         3        19

Residual standard error: 7.93603
Estimated effects may be unbalanced

R에서 기본으로 제공하는 명령어 aov를 이용하여 분산분석을 실시한다. 명령문에서 괄호 안은 분산분석을 위한 식, 사용할 자료명 순이다. ‘~’기호 앞의 변수가 종속변수, 기호 뒤의 변수는 요인이 된다. 위의 식은 교육방법(A, B, C, D)에 따른 시험성적의 평균 차이가 있는가를 알아보는 분석이다. 이 경우 사용하는 자료 ex.9 안에 식에 사용되는 변수 value와 method가 포함되어 있어야 한다.

### 결과 해석

In [4]:
anova(ex.10.aov)

Df Sum Sq    Mean Sq   F value  Pr(>F)    
method     3  712.5864 237.52881 3.771461 0.02804096
Residuals 19 1196.6310  62.98058       NA         NA

$\mu_1 = \mu_2 = \mu_3 = \mu_4 $를 각각의 교육방법에서 모집단의 평균성적이라 할 때, 검정하고자 하는 가설은 다음과 같다.

$H_0: \mu_1 = \mu_2 = \mu_3 = \mu_4$ vs $H_1 : H_0$ 는 사실이 아니다.

P값을 확인하기 위해서는 anova 명령어를 이용해야 한다. 앞의 결과에서 P값이 0.02804이므로 5% 유의수준 하에서 귀무가설은 기각된다. 즉, 주어진 자료에 의할 때 5% 유의수준 하에서 4가지 교육방법의 효과가 동일하다고 볼 수 없다.

In [5]:
summary(ex.10.aov)

            Df Sum Sq Mean Sq F value Pr(>F)  
method       3  712.6  237.53   3.771  0.028 *
Residuals   19 1196.6   62.98                 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

summary 명령어를 이용하는 방법도 있다.



## <예제 10-4> 자료에 대한 분산분석 과정
### 자료 입력

In [7]:
ex.10.two<-read.csv("ex10_4.csv",header=T)
ex.10.two

value treat block
1  1.7   A     a    
2  1.5   A     b    
3  0.1   A     c    
4  0.6   A     d    
5  3.4   B     a    
6  2.6   B     b    
7  2.3   B     c    
8  2.2   B     d    
9  2.3   C     a    
10 2.1   C     b    
11 0.8   C     c    
12 1.6   C     d

read.csv 명령어를 이용하여 ex10_2 파일을 ex.10.two 에 저장한다. 첫 번째 변수 value는 반응시간, 두 번째 변수 treat는 자극의 종류, 세 번째 변수 block은 사람으로 블록 효과이다. 

### 이원분류 분산분석

In [8]:
ex.10.two.aov<-aov(value~block+treat,data=ex.10.two)
ex.10.two.aov

Call:
   aov(formula = value ~ block + treat, data = ex.10.two)

Terms:
                   block    treat Residuals
Sum of Squares  3.480000 5.471667  0.455000
Deg. of Freedom        3        2         6

Residual standard error: 0.2753785
Estimated effects may be unbalanced

R에서 기본으로 제공하는 명령어 aov를 이용하여 분산분석을 실시한다. 명령문에서 괄호 안은 분산분석을 위한 식, 사용할 자료명 순이다. ‘~’기호 앞의 변수는 종속변수, 기호 뒤의 변수는 요인이 된다. 두 가지 이상의 요인이 필요한 경우 ‘+’ 기호를 사용한다. 위의 코드는 블록(block : 사람 1, 사람 2, 사람 3, 사람 4) 효과와 자극의 종류(treat : A, B, C)에 따른 반응시간의 처리 효과를 분석하는 코드로, 사용하는 자료 ex.10.two 안에 식에 사용되는 변수 value와 block, treat가 포함되어 있어야 한다.

### 결과 해석

In [9]:
anova(ex.10.two.aov)

Df Sum Sq   Mean Sq    F value  Pr(>F)      
block     3  3.480000 1.16000000 15.29670 0.0032315977
treat     2  5.471667 2.73583333 36.07692 0.0004524834
Residuals 6  0.455000 0.07583333       NA           NA

$\beta_1 = \beta_2 = \beta_3 = \beta_4$를 각각의 사람(여기서는 block 변수)에 따른 반응시간의 모평균이라 할 때, 검정하고자 하는 가설은 다음과 같다.  
$H_0: \beta_1 = \beta_2 = \beta_3 = \beta_4$ vs $H_1: H_0 $은 사실이 아니다. 


위의 분석결과에서 block의 효과에 대한 P값이 0.0032316으로 0.05보다 작으므로 귀무가설은 기각된다. 즉, 사람에 따른 반응시간의 기댓값이 동일하다고 할 수 없다.

In [10]:
summary(ex.10.two.aov)

            Df Sum Sq Mean Sq F value   Pr(>F)    
block        3  3.480  1.1600   15.30 0.003232 ** 
treat        2  5.472  2.7358   36.08 0.000452 ***
Residuals    6  0.455  0.0758                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1